### PR TITLE
This fixes #1 - Refactoring to inject agenda, added a few checks ...

### LINKED ISSUE #1: feedback on initial version
- [x] remove npm-debug.log from git
- [x] pass in agenda object instead of db connection
- [x] remove batchHash - needs to work across many servers
- [x] do a query in agenda for the batchKey to determine if job already exists - https://github.com/rschmukler/agenda#jobsmongoskin-query
- [x] remove define from this module.  Jobs will be defined on the agenda side

### PR DESCRIPTION
**Pull request by ecwillis**
... , and job is pulled from mongo to work with multiple machines

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -5,18 +5,23 @@ A batching tool for the agenda.js tool that will proxy data to a task that has n
 Initialize the library:
 
 ```javascript
+var Agenda = require('agenda');
 var Batch = require('batch');
-batch = new Batch('db-connection:24717');
+
+var agenda = new Agenda({ db: { address: 'db-connection:24717' } });
+batch = Batch( agenda );
```
Comment (jgallen23) [inline README.md]: `new Batch`

File: lib/batch.js [hunk 2]
```diff
@@ -4,51 +4,90 @@
...
   }
 
-  return init(dbConnection);
-
+  return init(agendaModule);
 }
 
-
 module.exports = function(dbConn) {
   return AgendaBatch(dbConn);
```
Comment (jgallen23) [inline lib/batch.js]: `new AgendaBatch`

File: lib/batch.js [hunk 3]
```diff
@@ -4,51 +4,90 @@
...
+      cb( null, job );
+
+    });
   }
 
   selfObj.cleanBatch = function( job ) {
-    if( selfObj.batchHash[ job.attrs.data.key ] !== undefined ) {
-      delete selfObj.batchHash[ job.attrs.data.key ];
-    }
+    job.remove(function(err, count) {});
```
Comment (jgallen23) [inline lib/batch.js]: what is this for?

File: lib/batch.js [hunk 4]
```diff
@@ -4,51 +4,90 @@
 
 var Agenda = require('agenda');
 
-var AgendaBatch = function(dbConnection) {
+var AgendaBatch = function(agendaModule) {
 
   var selfObj = {};
```
Comment (jgallen23) [inline lib/batch.js]: `var selfObj = this`

File: lib/batch.js [hunk 5]
```diff
@@ -4,51 +4,90 @@
...
 
-  var init = function(dbConnect) {
-    selfObj.agendaModule = new Agenda({ db: { address: dbConnection }});
+  var init = function(agendaMod) {
```
Comment (jgallen23) [inline lib/batch.js]: agendaMod is already defined above, you don't need to set it on the obj

File: lib/batch.js [hunk 6]
```diff
@@ -4,51 +4,90 @@
...
+
+      if ( jobs.length == 0 ) {
+        // No Jobs we're found... better make one up!
+        newJob = spawnNewJob(batchKey, data, processTime, jobName);
+        cb(null, newJob);
+
+        return;
+      }
+
+      if ( jobs.length > 1 ) {
```
Comment (jgallen23) [inline lib/batch.js:41]: I think you'd need to check if a job has been completed

File: lib/batch.js (hunk 3 above)
Comment (ecwillis) [inline lib/batch.js]: This was to ensure that the data was nuked. But, the new paradigm of job checking makes it un-necessary.

File: lib/batch.js [hunk 7]
```diff
@@ -2,53 +2,80 @@
...
+  selfObj.batch = function(batchKey, data, processTime, jobName, cb) {
+
+    if ( cb == undefined ) {
+      cb = function(err, job) {};
     }
 
-    selfObj.agendaModule.define(jobName, options, processor);
-  }
+    // REFACTOR
+    agendaModule.jobs( { 'data.key': batchKey, 'lastRunAt': { '$exists': false }  }, function(err, jobs) {
```
Comment (jgallen23) [inline lib/batch.js:27]: is there a running flag that you could check against too?  might simplify the logic below and guarantee we don't append to a running job

File: lib/batch.js [hunk 8]
```diff
@@ -2,53 +2,80 @@
...
+
+      cb( null, job );
+
+    });
+  }
+
+  return selfObj;
+}
 
 module.exports = function(dbConn) {
```
Comment (jgallen23) [inline lib/batch.js]: lets get rid of this (see above comment too)

File: lib/batch.js [hunk 9]
```diff
@@ -2,53 +2,80 @@
   AgendaBatch
 */
 
-var Agenda = require('agenda');
+var AgendaBatch = function(agendaModule) {
```
Comment (jgallen23) [inline lib/batch.js]: could just do `module.exports = function(...`

File: lib/batch.js (hunk 7 above)
Comment (ecwillis) [inline lib/batch.js:27]: There is... but it's relatively complex. In theory, simply checking to see if lastRunAt does not exist will filter out any job that has run or is running for any 'scheduled' task. The issue is with recurring tasks where lastRunAt exists but is scheduled to run again.

There is a method to check if a process is running (which is what is used below this) - though I have a feeling that most of the tasks you will be running are 'scheduled'.

Thoughts?

Comment (ecwillis) [inline lib/batch.js:27]: Nevermind - I answered my own question.

File: README.md (hunk 1 above)
Comment (jgallen23) [inline README.md]: still need to update this to new Batch(agenda)

**Pull request merged**